google/cloud/bigquery/dataset.py

```diff
@@ -382,6 +217,281 @@ def __repr__(self):
+
+    Examples:
+        >>> entry = AccessEntry('OWNER', 'userByEmail', 'user@example.com')
+
+        >>> view = {
+        ...     'projectId': 'my-project',
+        ...     'datasetId': 'my_dataset',
+        ...     'tableId': 'my_table'
+        ... }
+        >>> entry = AccessEntry(None, 'view', view)
+    """
+
+    def __init__(self, role=None, entity_type=None, entity_id=None):
+        self._properties = {}
+        if entity_type is not None:
+            self._properties[entity_type] = entity_id
+        self._properties["role"] = role
+        self._entity_type = entity_type
+
+    @property
+    def role(self) -> Optional[str]:
+        """The role of the entry."""
+        return self._properties.get("role")
+
+    @role.setter
+    def role(self, value):
+        self._properties["role"] = value
+
+    @property
+    def dataset(self) -> Optional[dict]:
+        """API resource representation of a dataset reference."""
+        return self._properties.get("dataset")
+
+    @dataset.setter
+    def dataset(self, value):
+        if self.role is not None:
+            raise ValueError(
+                "Role must be None for a dataset. Current " "role: %r" % (self.role)
+            )
+        if not isinstance(value, dict):
+            if isinstance(value, str):
+                value = DatasetReference.from_string(value)
+
+            if isinstance(value, (Dataset, DatasetListItem)):
+                value = value.reference
+
+            value = value.to_api_repr()
+
+        _helpers._set_sub_prop(self._properties, ["dataset", "dataset"], value)
+        _helpers._set_sub_prop(
+            self._properties,
+            ["dataset", "targetTypes"],
+            self._properties.get("targetTypes"),
+        )
+
+    @property
+    def target_types(self) -> Optional[List[str]]:
+        """Which resources that the dataset in this entry applies to."""
+        return _helpers._get_sub_prop(self._properties, ["dataset", "targetTypes"])
+
+    @target_types.setter
+    def target_types(self, value):
+        if "dataset" not in self._properties:
+            self._properties["dataset"] = {}
+        _helpers._set_sub_prop(self._properties, ["dataset", "targetTypes"], value)
+
+    @property
+    def routine(self) -> Optional[dict]:
+        """API resource representation of a routine reference."""
+        return self._properties.get("routine")
+
+    @routine.setter
+    def routine(self, value):
+        if self.role is not None:
+            raise ValueError(
+                "Role must be None for a routine. Current " "role: %r" % (self.role)
+            )
+        if not isinstance(value, dict):
+            if isinstance(value, str):
+                value = RoutineReference.from_string(value)
+
+            if isinstance(value, Routine):
+                value = value.reference
+
+            value = value.to_api_repr()
+        self._properties["routine"] = value
+
+    @property
+    def view(self) -> Optional[dict]:
```

Comment:
nit: we tend to use type `Dict` where we can.

```suggestion
    def view(self) -> Optional[Dict[str, Any]]:
```